dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/engine/ImplementableRuleService.java

```diff
@@ -77,7 +78,14 @@ public ImplementableRuleService( ProgramRuleService programRuleService )
 
         List<ProgramRule> programRulesByActionTypes = getProgramRulesByActionTypes( program, programStageUid );
 
-        getProgramRulesCache().put( program.getUid(), !programRulesByActionTypes.isEmpty() );
+        if ( programStageUid == null )
+        {
+            getProgramRulesCache().put( program.getUid(), !programRulesByActionTypes.isEmpty() );
+
+            return programRulesByActionTypes.stream().filter( rule -> rule.getProgramStage() == null )
```

Comment:
Could you maybe add a description of the logic here? We first cache whether a program has program rules with any program stage, and then return only rules with no program stage?

Reply:
Added.
At enrollment, only those rules should be sent for execution with are NOT associated with any ProgramStage.